main/contracts/index.ts

```diff
@@ -44,45 +29,16 @@ function parseAbi (abiData: string): Interface | undefined {
   }
 }
 
-// function scanEndpoint (contractAddress: Address, chainId: string) {
-//   if (chainId === '0x1') {
-//     return `https://api.etherscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=3SYU5MW5QK8RPCJV1XVICHWKT774993S24`
-//   } else if (chainId === '0x89') {
-//     return `https://api.polygonscan.com/api?module=contract&action=getsourcecode&address=0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270&apikey=YourApiKeyToken`
-//   } else if ('optimism') {
-//     return `https://api-optimistic.etherscan.io/api?module=contract&action=getsourcecode&address=0x80AA7cb0006d5DDD91cce684229Ac6e398864606&apikey=YourApiKeyToken`
-//   } else if ('arbitrum') {
-//     return `https://api.arbiscan.io/api?module=contract&action=getsourcecode&address=0x0000000000000000000000000000000000001004&apikey=YourApiKeyToken`
-//   }
-// }
-
-async function fetchSourceCode (contractAddress: Address, chainId: string) {
-  const res = await fetch(`https://api.etherscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=3SYU5MW5QK8RPCJV1XVICHWKT774993S24`)
-
-  if (res.status === 200 && (res.headers.get('content-type') || '').toLowerCase().includes('json')) {
-    const parsedResponse = await (res.json() as Promise<EtherscanSourceCodeResponse>)
-
-    if (parsedResponse.message === 'OK') return parsedResponse.result
-  }
-
-  return []
-}
-
-async function fetchAbi (contractAddress: Address, chainId: string): Promise<ContractSourceCodeResult | undefined> {
+export async function fetchContract (contractAddress: Address, chainId: string): Promise<ContractSource | undefined> {
   try {
-    const sources = await fetchSourceCode(contractAddress, chainId)
-
-    if (sources.length > 0) {
-      const source = sources[0]
-      const implementation = source.Implementation
-
-      if (implementation) {
-        // this is a proxy contract, return the ABI for the source
-        return fetchAbi(implementation, chainId)
-      }
-
-      return source
-    }
+    const [sourcifyContract, scanContract] = await Promise.all([fetchSourcifyContract(contractAddress, chainId), fetchScanContract(contractAddress, chainId)])
```

Comment:
the good part about creating a standard interface is that we can change the inputs and not have to update any of the implementation code. for example you might assign all the possible fetch functions when you import them and then this can look like:

`
const [contracts] = await Promise.all(fetches.map(fetchContract => fetchContract(contractAddress, chainId)))`

then you can loop through and pick the first one, with the ability to add new sources elsewhere and never change this code

Reply:
This is all worth doing though I would argue that YAGNI until there is a third source.

Reply:
I don't think its overengineering at all. there are numerous benefits to creating a clean contract between components beyond saving future developers refactoring time, though that is also a benefit. it promotes readable and testable code through proper encapsulation of implementation details. it looks like your latest changes went a long towards addressing it though

Reply:
I don't disagree that it's beneficial and the right way to go.  Just that for only two sources, I would personally punt the refactor to when / if a third is introduced.